dali/python/nvidia/dali/plugin/pytorch/experimental/proxy/__init__.py

```diff
@@ -0,0 +1,771 @@
+                        dataset,
+                        batch_size=batch_size,
+                        num_workers=nworkers,
+                        drop_last=True,
+                    )
+
+                    for data, target in loader:
+                        # consume it
+
+        Example 2 - Manual execution using DALI proxy / DALI server and PyTorch's default_collate:
+
+            .. code-block:: python
+
+                @pipeline_def
+                def my_pipe():
+                    a = fn.external_source(name="a", no_copy=True)
+                    b = fn.external_source(name="b", no_copy=True)
+                    return a + b, a - b
+
+                with dali_proxy.DALIServer(
+                    my_pipe(device='cpu', batch_size=batch_size,
+                            num_threads=3, device_id=None)) as dali_server:
+
+                    outs = []
+                    for _ in range(batch_size):
+                        a = np.array(np.random.rand(3, 3), dtype=np.float32)
+                        b = np.array(np.random.rand(3, 3), dtype=np.float32)
+
+                        if named_arguments:
+                            out0, out1 = dali_server.proxy(a=a, b=b)
+                        else:
+                            out0, out1 = dali_server.proxy(a, b)
+
+                        outs.append((a, b, out0, out1))
+
+                    outs = torch.utils.data.dataloader.default_collate(outs)
+
+                    a, b, a_plus_b, a_minus_b = dali_server.produce_data(outs)
+
+        Example 3 - Full integration with PyTorch but using the original PyTorch DataLoader
+
+            .. code-block:: python
+
+                pipe = rn50_train_pipe(...)
+                with dali_proxy.DALIServer(pipe) as dali_server:
+                    dataset = torchvision.datasets.ImageFolder(
+                        jpeg, transform=dali_server.proxy, loader=read_filepath)
+
+                    # Using PyTorch DataLoader directly
+                    loader = torch.utils.data.DataLoader(
+                        dataset,
+                        batch_size=batch_size,
+                        num_workers=nworkers,
+                        drop_last=True,
+                    )
+
+                    for data, target in loader:
+                        # replaces the output reference with actual data
+                        data = dali_server.produce_data(data)
+                        ...
+        """
+        if not isinstance(pipeline, _Pipeline):
+            raise TypeError(f"Expected an NVIDIA DALI pipeline, got: {pipeline}")
+        else:
+            self._pipe = pipeline
+
+        # get the dali pipeline input names
+        self._dali_input_names = _external_source_node_names(self._pipe)
+        if len(self._dali_input_names) == 0:
+            raise RuntimeError("The provided pipeline doesn't have any inputs")
+
+        parameters = [Parameter("self", Parameter.POSITIONAL_OR_KEYWORD)]
```

Comment:
I think it's a dangerous to have positional inputs.
`_external_source_node_names` is populated with `pipe._ops` which, in turn, is populated via graph traversal from outputs to inputs. The result is that the pipelines below will have their outputs in different order:
```Python
@dali.pipeline_def(batch_size=1, num_threads=1)
def pipe_a():
    a = fn.external_source(name="a")
    b = fn.external_source(name="b")
    return a + b, b - a

@dali.pipeline_def(batch_size=1, num_threads=1)
def pipe_b():
    a = fn.external_source(name="a")
    b = fn.external_source(name="b")
    return b - a, a + b

p = pipe_a()
p.build()
print(_external_source_node_names(p))
q = pipe_b()
q.build()
print(_external_source_node_names(q))
```
Output:
```
['a', 'b']
['b', 'a']
```

Reply:
disabled positional inputs if more than 1